lib/compiler.js

```diff
@@ -598,3 +604,67 @@ function resolveAppScriptPath(rootDir, relativePath, resolveOptions) {
   var fixedFile = fixFileExtension(resolvedPath, files, false);
   return (fixedFile === undefined ? resolvedPath : fixedFile);
 }
+
+function buildAllMixinInstructions(appRootDir, mixinSources, options) {
+  var extensions = _.without(_.keys(require.extensions),
+    _.keys(getExcludedExtensions()));
+  var files = options.mixins || [];
+
+  mixinSources.forEach(function(dir) {
+    dir = tryResolveAppPath(appRootDir, dir);
+    if (!dir) {
+      debug('Skipping unknown module source dir %j', dir);
+      return;
+    }
+    files = files.concat(findScripts(dir, extensions));
+  });
+
+  var mixins = {};
+  files.forEach(function(filepath) {
+    var dir = path.dirname(filepath);
+    var ext = path.extname(filepath);
+    var name = path.basename(filepath, ext);
+    var metafile = path.join(dir, name + FILE_EXTENSION_JSON);
+    name = normalizeMixinName(name, options);
+    var meta = {};
+    meta.name = name;
+    if (fs.existsSync(metafile)) {
+      // May overwrite name, not id or filepath
+      _.extend(meta, require(metafile));
+    }
+    meta.filepath = filepath;
+    mixins[meta.name] = meta;
+  });
+
+  return mixins;
```

Comment:
The instructions entity should be an array of mixins to create, not a lookup table. That is the convention we use elsewhere, see model instructions for an example.